src/xlog/xlog.3
.\"(c) Copyright 1992, 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms
.\"and conditions for redistribution.
.\"
.\" $Id: xlog.3,v 1.2 2003/06/17 05:10:56 seth Exp $
.TH XLOG 3X "15 June 1993"
xlog_parms, xlog_create, xlog_destroy, xlog_write, xlog_control -- general purpose logging facility
.SH SYNOPSIS
.LP
.nf
.ft B
#include "xlog.h"
.LP
.ft B
xlog_h xlog_create( type, id, flags, ... )
xlog_e type ;
char *id ;
int flags ;
.LP
.ft B
int xlog_parms( type, ... )
xlog_e type ;
.LP
.ft B
void xlog_destroy( xlog )
xlog_h xlog ;
.LP
.ft B
void xlog_write( xlog, buf, len, flags, ... )
xlog_h xlog ;
char buf[] ;
int len ;
int flags ;
.LP
.ft B
int xlog_control( xlog, cmd, ... )
xlog_h xlog ;
xlog_cmd_e cmd ;
.SH DESCRIPTION
The purpose of this library is to provide a general purpose logging facility
by providing
.I xlogs,
logging objects that may be connected to various existent logging facilities.
Currently, the only logging facilities supported are
.I "syslog(3)"
and logging to files.
Log entries are timestamped lines which may contain arbitrary information.
.\" ********************* xlog_create ***********************
.LP
.B xlog_create()
creates a new xlog of the specified
.I type.
Possible type values are:
.RS
.TP 20
.SB XLOG_SYSLOG
Varargs: \fIint facility, int priority\fP.
The xlog will be connected to
.I "syslog(3)."
.I facility
determines the syslog facility to use for logged messages and
.I priority
is the default message priority.
.TP
.SB XLOG_FILELOG
Varargs: \fIchar *pathname, int flags [, int flags]\fP.
The xlog will be connected to the file identified by
.I pathname.
The variable part of the argument list has the same semantics as the
argument list of the
.I "open(2)"
system call.
.RE
.LP
All xlogs have an id, specified by the
.I id
argument. The
.I flags
argument is formed by ORing one or more of the following constants:
.RS
.TP 20
.SB XLOG_NO_ERRNO
do not replace
.I "%m"
with an explanation of the current value of errno.
.TP
.SB XLOG_NO_SIZECHECK
.I "(XLOG_FILELOG only)"
do not perform size checks on the file.
.TP
.SB XLOG_PRINT_ID
precede each log entry with the xlog id
.TP
.SB XLOG_PRINT_PID
precede each log entry with the process id
(the process id will follow the xlog id)
.RE
.LP
Flags that do not apply to the xlog are ignored.
The contant
.SM XLOG_NOFLAGS
can be used if you don't want to specify any flags.
.\" ********************* xlog_parms ***********************
.LP
.B xlog_parms()
sets default parameters for the specified xlog
.I type:
.RS
.TP 20
.SB XLOG_SYSLOG
3 arguments are expected which correspond one-to-one to the arguments of
.I "openlog(3)."
The defaults, in case this function is not used, are:
"XLOG", LOG_PID + LOG_NOWAIT, LOG_USER
.TP
.SB XLOG_FILELOG
No action.
.RE
.LP
.B xlog_parms()
should be invoked before any xlogs of the specified type
are created.
.\" ********************* xlog_destroy ***********************
.LP
.B xlog_destroy()
destroys an xlog. The action taken depends on the type of the xlog:
.RS
.TP 20
.SB XLOG_SYSLOG
if this is the last xlog using syslog, then
.I "closelog(3)"
is invoked.
.TP
.SB XLOG_FILELOG
The file is closed.
.RE
.\" ********************* xlog_control ***********************
.LP
.B xlog_control()
applies control operations to an xlog. Certain operations are common to
all xlogs while others are type-specific. The common ones are:
.RS
.TP 15
.SB XLOG_LINK
Argument list: \fIxlog_h link_to\fP.
Link the specified xlog to the one provided as argument.
(if the argument is
.SM NULL
any existent link is severed).
Linking xlogs has the effect that if one finds an error it uses the
other to report it.
.TP
.SB XLOG_CALLBACK
Argument list: \fIvoid (*callback)(), void *arg\fP.
This function will be invoked in case of error while writing a log
entry. It will be given
3 arguments: the xlog handle, an integer that indicates the type
of error and the
.I arg
specified in this call. Possible errors include:
.RS
.TP 15
.SB XLOG_ENOMEM
lack of memory
.TP
.SB XLOG_EOPEN
.I "open(2)"
failed
.TP
.SB XLOG_EFSTAT
.I "fstat(2)"
failed
.TP
.SB XLOG_ESIZE
hard limit exceeded
.RE
.TP
.SB XLOG_SETFLAG
Argument list: \fIint flag, int *value\fP.
The value of
.I flag
(one of those listed before) is set according to
.I "*value"
which should be either 0 or 1.
The old flag value is placed in
.I "*value."
.TP
.SB XLOG_GETFLAG
Argument list: \fIint flag, int *value\fP.
The value of
.I flag
(one of those listed before) is placed in
.I "*value."
.RE
.LP
The type-specific operations have the following semantics when the
operation is applied to an incompatible xlog: if the operation returns
a result to the caller (for example, check \fPXLOG_GETFD\fP), the
operation will be considered to have failed, otherwise the operation
will be considered successful.
.LP
Xlogs of type
.B XLOG_SYSLOG
also support the following operations:
.RS
.TP 15
.SB XLOG_FACILITY
Argument list: \fIint facility\fP.
Sets the syslog facility to the specified value.
.TP
.SB XLOG_LEVEL
Argument list: \fIint level\fP.
Sets the default syslog level for this xlog to the specified value.
.TP
.SB XLOG_PREEXEC
Argument list: \fIvoid\fP.
Prepares the xlog for an impending exec operation
.TP
.SB XLOG_POSTEXEC
Argument list: \fIvoid\fP.
Informs the xlog that the exec failed
.RE
.LP
Xlogs of type
.B XLOG_FILELOG
also support the following operations:
.RS
.TP 15
.SB XLOG_LIMITS
Argument list: \fIunsigned soft_limit, unsigned hard_limit\fP.
Sets soft and hard limits on the size of the file.
When any of the limits is exceeded a message is sent to the linked xlog.
(if there is no linked xlog, no message is sent)
When the soft limit is exceeded a warning message is sent to the linked xlog
(if the linked xlog is of the
.SB XLOG_SYSLOG
type, the message will be sent at the
.I LOG_ALERT
level).
If the exceeded limit is the hard limit, logging is terminated.
The actual file size is checked every time this operation is applied to
the file.
If logging was terminated because the hard limit was exceeded and
this operation increases the hard limit beyond the actual file size,
logging will be resumed.
.TP
.SB XLOG_SIZECHECK
Argument list: \fIvoid\fP.
Checks the actual file size.
.TP
.SB XLOG_GETFD
Argument list: \fIint *value\fP.
Places in
.I "*value"
the file descriptor of the log file.
.RE
.\" ********************* xlog_write ***********************
.LP
.B xlog_write()
writes a message to the specified xlog. A
.SM NEWLINE
is always appended to the message.
The first occurrence of "%m" in
.I buf
is replaced by a string explaining the current value of
.I errno.
The
.I flags
argument is formed in the same way as in
.B xlog_create().
One additional constant is available:
.RS
.TP 20
.SB XLOG_SET_LEVEL
.I "(XLOG_SYSLOG only)"
the next argument is an integer that should be used as the syslog level
for this message instead of the default level of the xlog.
.RE
.SH "RETURN VALUES"
.B xlog_create()
returns an xlog handle or
.SM NULL
if it fails.
.LP
.B xlog_control()
returns an error code (it returns
.SM XLOG_ENOERROR
if it is successful).
.LP
.B xlog_parms()
returns an error code (it returns
.SM XLOG_ENOERROR
if it is successful).
.SH "SEE ALSO"
openlog(3), syslog(3), closelog(3)
.SH BUGS
.LP
Only the first occurrence of
.I "%m"
is replaced by an errno explanation.
.LP
There is no check for cycles when linking xlogs. In particular it is
possible to link a xlog to itself.


ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive
	operations on empty hashes.

2002-05-13  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): Shared libfsma no longer
	breaks with libtool (on both linux and bsd).

2002-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): back out shared libfsma;
	this breaks libtool in a strange way that requires investigation

2002-05-10  James Tanis  <jtt@winwood.sysdetect.com>

	* src/fsma/fsma.c: Formatting and simplify coalesce code.

	* src/fsma/Makefile (BK_LARGE_LIB): Make libfsma shared.

	* src/dict/bst.c (bst_redelete): Improve efficiency by avoiding
	unnecessary checks and recusive calls.

2002-02-22  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_NUKE): delete contents and dict itself
	(_DICT_NUKE_WORK_LOOP): common code for DICT_NUKE, DICT_NUKE_CONTENTS

	* src/dict/dictimpl.h (HANDLE_ERROR): Eliminate unused id argument
	[use BK_FUNCNAME  here instead if you need it for debugging]

2002-01-31  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/Makefile (BK_LARGE_LIB): use ${LIBEXT} for libtool build

2002-01-09  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/fsma.c: INSIGHT->INSURE


src/xlog/xlog.c
/*
 * (c) Copyright 1992, 1993 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms
 * and conditions for redistribution.
 */

static const char version[] = VERSION;

#include <stdarg.h>

#include "xlog.h"
#include "impl.h"
#include <stdlib.h>
#include "clchack.h"

extern struct xlog_ops __xlog_filelog_ops ;
#ifndef NO_SYSLOG
extern struct xlog_ops __xlog_syslog_ops ;
#endif

struct lookup_table
{
	struct xlog_ops	*ops ;
	xlog_e				type ;
} ;

static struct lookup_table ops_lookup_table[] =
	{
		{ &__xlog_filelog_ops,				XLOG_FILELOG	},
#ifndef NO_SYSLOG
		{ &__xlog_syslog_ops,				XLOG_SYSLOG		},
#endif
		{ NULL, 0 }
	} ;


#define CALLBACK( xp, status ) 																\
		if ( (xp)->xl_callback ) 																\
			(*(xp)->xl_callback)( (xlog_h)(xp), status, (xp)->xl_callback_arg )


PRIVATE void xlog_link(xlog_s *client, xlog_s *server) ;
PRIVATE void xlog_unlink(xlog_s *xp) ;


PRIVATE struct xlog_ops *xlog_ops_lookup(register xlog_e type)
{
	register struct lookup_table *ltp ;

	for ( ltp = &ops_lookup_table[ 0 ] ; ltp->ops ; ltp++ )
		if ( ltp->type == type )
			break ;
	return( ltp->ops ) ;
}



/* VARARGS3 */
xlog_h xlog_create( xlog_e type, char *id, int flags, ... )
{
	xlog_s				*xp ;
	va_list				ap ;
	struct xlog_ops	*xops ;
	int					status ;

	if ( ( xp = NEW( xlog_s ) ) == NULL )
		return( NULL ) ;

	if ( id == NULL || ( xp->xl_id = __xlog_new_string( id ) ) == NULL )
	{
		FREE( xp ) ;
		return( NULL ) ;
	}

	xops = xlog_ops_lookup( type ) ;

	if ( xops != NULL )
	{
		va_start( ap , flags ) ;
		xp->xl_ops = xops ;
		status = XL_INIT( xp, ap ) ;
		va_end( ap ) ;

		if ( status == XLOG_ENOERROR )
		{
			xp->xl_flags = flags ;
			xp->xl_type = type ;
			xp->xl_clients = XLOG_NULL ;
			xp->xl_use = XLOG_NULL ;
			return( (xlog_h) xp ) ;
		}
	}

	free( xp->xl_id ) ;
	FREE( xp ) ;
	return( NULL ) ;
}



PRIVATE void xlog_link(xlog_s *client, xlog_s *server)
{
	client->xl_use = server ;
	if ( server == NULL )
		return ;

	if ( server->xl_clients == XLOG_NULL )
	{
		INIT_LINKS( client, xl_other_users ) ;
		server->xl_clients = client ;
	}
	else
		LINK( server, client, xl_other_users ) ;
}


PRIVATE void xlog_unlink(xlog_s *xp)
{
	xlog_s *server = xp->xl_use ;

	/*
	 * Step 1: remove from server chain
	 */
	if ( server != XLOG_NULL )
	{
		if ( server->xl_clients == xp )
			if ( NEXT( xp, xl_other_users ) == xp )
				server->xl_clients = XLOG_NULL ;
			else
				server->xl_clients = NEXT( xp, xl_other_users ) ;
		else
			UNLINK( xp, xl_other_users ) ;
	}

	/*
	 * Step 2: If we have users, clear their link to us.
	 */
	if ( xp->xl_clients != NULL )
	{
		xlog_s *xp2 = xp->xl_clients ;

		do
		{
			xp2->xl_use = XLOG_NULL ;
			xp2 = NEXT( xp2, xl_other_users ) ;
		}
		while ( xp2 != xp->xl_clients ) ;
	}
}


PRIVATE void xlog_flags(xlog_s *xp, xlog_cmd_e cmd, va_list ap)
{
	int	flag			= va_arg( ap, int ) ;
	int	old_value	= ( ( xp->xl_flags & flag ) != 0 ) ;
	int	*valp			= va_arg( ap, int * ) ;

	if ( cmd == XLOG_SETFLAG )
	{
		if ( *valp )
			xp->xl_flags |= flag ;
		else
			xp->xl_flags &= ~flag ;
	}
	*valp = old_value ;
}


void xlog_destroy(xlog_h xlog)
{
	xlog_s *xp = XP( xlog ) ;

	xlog_unlink( xp ) ;
	XL_FINI( xp ) ;
	free( xp->xl_id ) ;
	FREE( xp ) ;
}


/* VARARGS4 */
void xlog_write( xlog_h xlog, char buf[], int len, int flags, ... )
{
	xlog_s	*xp = XP( xlog ) ;
	va_list	ap ;
	int		status ;

	va_start( ap, flags ) ;
	status = XL_WRITE( xp, buf, len, flags, ap ) ;
	va_end( ap ) ;

	if ( status != XLOG_ENOERROR )
	{
		CALLBACK( xp, status ) ;
	}
}


/* VARARGS2 */
int xlog_control( xlog_h xlog, xlog_cmd_e cmd, ... )
{
	va_list	ap ;
	xlog_s	*xp		= XP( xlog ) ;
	int		status	= XLOG_ENOERROR ;

	va_start( ap, cmd ) ;

	switch ( cmd )
	{
		case XLOG_LINK:
			xlog_unlink( xp ) ;
			xlog_link( xp, va_arg( ap, xlog_s * ) ) ;
			xp->xl_callback_arg = va_arg( ap, void * ) ;
			break ;

		case XLOG_CALLBACK:
			xp->xl_callback = va_arg( ap, voidfunc ) ;
			break ;

		case XLOG_GETFLAG:
		case XLOG_SETFLAG:
			xlog_flags( xp, cmd, ap ) ;
			break ;

		/*
		 * If the 'cmd' is not supported by the underlying logging object,
		 * XL_CONTROL should return XLOG_EBADOP, if the 'cmd' returns
		 * information to the caller (for example, XLOG_GETFD returns
		 * a file descriptor, so if this command is applied to a xlog
		 * that doesn't support it, it should return XLOG_EBADOP).
		 * XL_CONTROL should return XLOG_ENOERROR otherwise.
		 */
		default:
			status = XL_CONTROL( xp, cmd, ap ) ;
	}

	va_end( ap ) ;

	return( status ) ;
}


int xlog_parms( xlog_e type, ... )
{
	va_list	ap ;
	int		status ;

	va_start( ap, type ) ;
	switch ( type )
	{
#ifndef NO_SYSLOG
		case XLOG_SYSLOG:
			status = (*__xlog_syslog_ops.parms)( ap ) ;
			break ;
#endif
		case XLOG_FILELOG:
			status = (*__xlog_filelog_ops.parms)( ap ) ;
			break ;

		default:
			status = XLOG_ENOERROR ;
	}
	va_end( ap ) ;
	return( status ) ;
}

